File: internal/tools/glob.go
package tools

import (
	"fmt"
	"io/fs"
	"os"
	"path"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
)

type GlobTool struct{}

func (GlobTool) Name() string {
	return "glob"
}

func (GlobTool) Description() string {
	return "Find files matching a glob pattern. Returns matching file paths sorted alphabetically. Useful for discovering files by name or extension."
}

func (GlobTool) Parameters() map[string]any {
	return objectSchema([]string{"pattern"}, map[string]any{
		"pattern": stringSchema("The glob pattern to match files against"),
		"path":    stringSchema("Directory to search in (defaults to cwd)"),
	})
}

func (GlobTool) Call(input map[string]any) Result {
	pattern, err := stringArg(input, "pattern")
	if err != nil {
		return Result{Output: "Error: " + err.Error(), IsError: true}
	}
	searchPath, ok, err := optionalStringArg(input, "path")
	if err != nil {
		return Result{Output: "Error: " + err.Error(), IsError: true}
	}
	if !ok || searchPath == "" {
		searchPath, err = os.Getwd()
		if err != nil {
			return Result{Output: "Error: " + err.Error(), IsError: true}
		}
	}

	files, err := expandGlob(pattern, searchPath)
	if err != nil {
		return Result{Output: "Error: " + err.Error(), IsError: true}
	}
	sort.Strings(files)
	if len(files) == 0 {
		return Result{Output: "No files matched the pattern."}
	}
	return Result{Output: strings.Join(files, "\n")}
}

func (GlobTool) RenderToolCall(input map[string]any) string {
	pattern, _ := stringArg(input, "pattern")
	searchPath, ok, _ := optionalStringArg(input, "path")
	description := toolName("glob") + " " + fileName(pattern)
	if ok && searchPath != "" {
		description += muted(" in " + searchPath)
	}
	return description
}

func (GlobTool) RenderResult(result Result) string {
	if result.IsError {
		return failure(result.Output)
	}
	return truncateLines(result.Output, 20, "more files")
}

func expandGlob(pattern, searchPath string) ([]string, error) {
	if !strings.Contains(pattern, "**") {
		matches, err := filepath.Glob(resolvePattern(pattern, searchPath))
		if err != nil {
			return nil, err
		}
		return onlyFiles(matches), nil
	}

	root, absolutePattern, err := globRoot(pattern, searchPath)
	if err != nil {
		return nil, err
	}

	patternSlash := filepath.ToSlash(absolutePattern)
	if runtime.GOOS == "windows" {
		patternSlash = strings.ToLower(patternSlash)
	}

	var matches []string
	err = filepath.WalkDir(root, func(current string, entry fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return nil
		}
		if entry.IsDir() {
			return nil
		}
		candidate := filepath.ToSlash(current)
		compare := candidate
		if runtime.GOOS == "windows" {
			compare = strings.ToLower(compare)
		}
		if matchGlobSegments(patternSlash, compare) {
			abs, err := filepath.Abs(current)
			if err != nil {
				return nil
			}
			matches = append(matches, abs)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	return matches, nil
}

func resolvePattern(pattern, searchPath string) string {
	if filepath.IsAbs(pattern) {
		return pattern
	}
	return filepath.Join(searchPath, pattern)
}

func onlyFiles(paths []string) []string {
	files := make([]string, 0, len(paths))
	for _, item := range paths {
		info, err := os.Stat(item)
		if err != nil || info.IsDir() {
			continue
		}
		abs, err := filepath.Abs(item)
		if err != nil {
			continue
		}
		files = append(files, abs)
	}
	return files
}

func globRoot(pattern, searchPath string) (string, string, error) {
	absolutePattern := resolvePattern(pattern, searchPath)
	cleanPattern := filepath.Clean(absolutePattern)
	parts := strings.Split(filepath.ToSlash(cleanPattern), "/")

	rootParts := []string{}
	for _, part := range parts {
		if hasGlobMeta(part) {
			break
		}
		rootParts = append(rootParts, part)
	}

	root := strings.Join(rootParts, "/")
	if root == "" {
		root = string(filepath.Separator)
	}
	if runtime.GOOS != "windows" && strings.HasPrefix(cleanPattern, string(filepath.Separator)) && !strings.HasPrefix(root, "/") {
		root = "/" + root
	}

	info, err := os.Stat(filepath.FromSlash(root))
	if err != nil {
		return "", "", err
	}
	if !info.IsDir() {
		return "", "", fmt.Errorf("%s is not a directory", root)
	}
	return filepath.FromSlash(root), cleanPattern, nil
}

func hasGlobMeta(value string) bool {
	return strings.ContainsAny(value, "*?[")
}

func matchGlobSegments(patternValue, candidateValue string) bool {
	patternParts := splitPath(patternValue)
	candidateParts := splitPath(candidateValue)
	return matchParts(patternParts, candidateParts)
}

func splitPath(value string) []string {
	value = strings.Trim(value, "/")
	if value == "" {
		return nil
	}
	return strings.Split(value, "/")
}

func matchParts(patternParts, candidateParts []string) bool {
	if len(patternParts) == 0 {
		return len(candidateParts) == 0
	}
	if patternParts[0] == "**" {
		if matchParts(patternParts[1:], candidateParts) {
			return true
		}
		for i := range candidateParts {
			if matchParts(patternParts[1:], candidateParts[i+1:]) {
				return true
			}
		}
		return false
	}
	if len(candidateParts) == 0 {
		return false
	}
	ok, err := path.Match(patternParts[0], candidateParts[0])
	if err != nil || !ok {
		return false
	}
	return matchParts(patternParts[1:], candidateParts[1:])
}

File: kai/tools/glob_.py
from __future__ import annotations

import fnmatch
import os
import pathlib
from typing import Any

from kai.tools.base import (
    Result,
    Tool,
    object_schema,
    optional_string_arg,
    string_schema,
)
from kai.tools.colors import failure, file_name, muted, tool_name
from kai.tools.format import truncate_lines


class GlobTool(Tool):
    def name(self) -> str:
        return "glob"

    def description(self) -> str:
        return "Find files matching a glob pattern. Returns matching file paths sorted alphabetically. Useful for discovering files by name or extension."

    def parameters(self) -> dict[str, Any]:
        return object_schema(
            ["pattern"],
            {
                "pattern": string_schema("The glob pattern to match files against"),
                "path": string_schema("Directory to search in (defaults to cwd)"),
            },
        )

    def call(self, input: dict[str, Any]) -> Result:
        pattern = input.get("pattern")
        if not isinstance(pattern, str) or not pattern:
            return Result(
                output="Error: missing required argument 'pattern'", is_error=True
            )

        search_path, path_specified = optional_string_arg(input, "path")
        if not path_specified or not search_path:
            try:
                search_path = os.getcwd()
            except OSError as e:
                return Result(output=f"Error: {e}", is_error=True)

        try:
            files = expand_glob(pattern, search_path)
        except Exception as e:
            return Result(output=f"Error: {e}", is_error=True)

        files.sort()
        if not files:
            return Result(output="No files matched the pattern.")
        return Result(output="\n".join(files))

    def render_tool_call(self, input: dict[str, Any]) -> str:
        pattern = input.get("pattern", "")
        search_path, path_specified = optional_string_arg(input, "path")
        description = f"{tool_name('glob')} {file_name(pattern)}"
        if path_specified and search_path:
            description += muted(f" in {search_path}")
        return description

    def render_result(self, result: Result) -> str:
        if result.is_error:
            return failure(result.output)
        return truncate_lines(result.output, 20, "more files")


def expand_glob(pattern: str, search_path: str) -> list[str]:
    if "**" not in pattern:
        matched = fnmatch.filter(
            _walk_files(search_path),
            os.path.join(search_path, pattern)
            if not os.path.isabs(pattern)
            else pattern,
        )
        return _only_files(matched)

    root, abs_pattern = _glob_root(pattern, search_path)
    abs_root = os.path.abspath(root) if not os.path.isabs(root) else root

    matches: list[str] = []
    pattern_parts = _split_path(abs_pattern)
    for dirpath, _dirnames, filenames in os.walk(abs_root):
        for filename in filenames:
            candidate = os.path.join(dirpath, filename)
            candidate_parts = _split_path(candidate)
            if _match_parts(pattern_parts, candidate_parts):
                matches.append(os.path.abspath(candidate))
    return matches


def _walk_files(search_path: str) -> list[str]:
    files: list[str] = []
    for dirpath, _dirnames, filenames in os.walk(search_path):
        for filename in filenames:
            files.append(os.path.join(dirpath, filename))
    return files


def _only_files(paths: list[str]) -> list[str]:
    result: list[str] = []
    for item in paths:
        try:
            if os.path.isfile(item):
                result.append(os.path.abspath(item))
        except OSError:
            continue
    return result


def _glob_root(pattern: str, search_path: str) -> tuple[str, str]:
    if os.path.isabs(pattern):
        abs_pattern = pattern
    else:
        abs_pattern = os.path.join(search_path, pattern)
    abs_pattern = os.path.normpath(abs_pattern)
    parts = _split_path(abs_pattern)
    root_parts: list[str] = []
    for part in parts:
        if _has_glob_meta(part):
            break
        root_parts.append(part)
    if os.name == "nt" and len(root_parts) >= 1 and root_parts[0].endswith(":"):
        root = root_parts[0] + os.sep
        if len(root_parts) > 1:
            root = os.path.join(root, *root_parts[1:])
    else:
        root = (
            os.sep + os.path.join(*root_parts)
            if abs_pattern.startswith(os.sep)
            else os.path.join(*root_parts)
        )
        if not root:
            root = os.sep

    if not os.path.isdir(root):
        root = os.path.dirname(root)

    return root, abs_pattern


def _has_glob_meta(value: str) -> bool:
    return any(c in value for c in "*?[")


def _split_path(value: str) -> list[str]:
    parts = pathlib.PurePosixPath(value).parts
    if os.sep == "\\":
        parts = pathlib.PureWindowsPath(value).parts
    result = [p for p in parts if p not in ("/", "\\", "")]
    return result


def _match_parts(pattern_parts: list[str], candidate_parts: list[str]) -> bool:
    if not pattern_parts:
        return not candidate_parts
    if pattern_parts[0] == "**":
        if _match_parts(pattern_parts[1:], candidate_parts):
            return True
        for i in range(len(candidate_parts)):
            if _match_parts(pattern_parts[1:], candidate_parts[i + 1 :]):
                return True
        return False
    if not candidate_parts:
        return False
    if not fnmatch.fnmatch(candidate_parts[0], pattern_parts[0]):
        return False
    return _match_parts(pattern_parts[1:], candidate_parts[1:])
